tests/unit/instrument_malloc.hpp

```diff
@@ -0,0 +1,138 @@
+#pragma once
+
+// Base class for scoped-instrumentation of glibc malloc.
+//
+// For the lifetime of a `with_instrumented_malloc` object,
+// global memory allocation hooks will be set so that
+// the virtual `on_malloc`, `on_realloc`, `on_memalign`
+// and `on_free` calls will be invoked before the corresponding
+// `malloc`, `realloc` etc. is executed.
+//
+// Scopes of `with_instrumented_malloc` may be nested, but:
+//   * Don't interleave lifetimes of these objects and expect things
+//     to work!
+//   * Don't try and create new `with_instrumented_malloc` instances
+//     from within an `on_malloc` callback (or others).
+//   * Definitely don't try and use this in a multithreaded context.
+//
+// Calling code should check CAN_INSTRUMENT_MALLOC preprocessor
+// symbol to see if this functionality is available.
+
+#include <cstddef>
+
+#if (__GLIBC__==2)
+#include <malloc.h>
+#define CAN_INSTRUMENT_MALLOC
+#endif
+
+namespace testing {
+
+#ifdef CAN_INSTRUMENT_MALLOC
```

Comment:
It's there, because it is the only way available to test the padding part of the padded allocator; without it, we'd have to use an external tool or the address sanitizer to check accesses to the padded region.